### PR TITLE
feat: example escrow compressed tokens with pda 

### PR DESCRIPTION
**Pull request by ananas-block**
# Changes:
- add example directory
- add token escrow example anchor program
  - it has two instructions
    1. escrow compressed tokens and create a timelock pda
    2. withdraw compressed tokens provided the timelock expired
- marked a couple cpi signer accounts in account compression, compressed pda and token programs at not mutable since it is not required and destroys parallelism with writelocks on general utility accounts

### CODE REVIEW

File: examples/token-escrow/programs/token-escrow/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        input_token_data_with_context,
+        output_compressed_accounts,
+        output_state_merkle_tree_account_indices,
+        pubkey_array,
+        is_compress: false,
+        compression_amount: None,
+    };
+
+    let mut inputs = Vec::new();
+    CompressedTokenInstructionDataTransfer::serialize(&inputs_struct, &mut inputs).unwrap();
```
Comment (vadorovsky) [inline examples/token-escrow/programs/token-escrow/src/lib.rs:187]:
```suggestion
    let inputs = input_struct.try_to_vec()?;
```

But yeah, this `serialize` method is exactly what I wanted to use for the "account buffer" idea. :sweat_smile: Here it's not necessary though.

File: programs/compressed-token/src/process_transfer.rs [hunk 2]
```diff
@@ -482,6 +486,113 @@ pub mod transfer_sdk {
...
+            owner_if_delegate_is_signer,
+            output_compressed_accounts,
+            root_indices,
+            proof,
+            mint,
+            is_compress,
+            compression_amount,
+        );
+        let mut inputs = Vec::new();
+        CompressedTokenInstructionDataTransfer::serialize(&inputs_struct, &mut inputs).unwrap();
```
Comment (vadorovsky) [inline programs/compressed-token/src/process_transfer.rs:504]:
```suggestion
        let inputs = inputs_struct.try_to_vec()?;
```

File: test-utils/src/test_indexer.rs [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+        &mut self,
+        event: PublicTransactionEvent,
+    ) -> Vec<usize> {
+        for compressed_account in event.input_compressed_accounts.iter() {
+            let index = self
+                .compressed_accounts
+                .iter()
+                .position(|x| x.compressed_account == compressed_account.compressed_account)
+                .expect("compressed_account not found");
+            self.compressed_accounts.remove(index);
```
Comment (vadorovsky) [inline test-utils/src/test_indexer.rs:175]:
```suggestion
            self.compressed_accounts.retain(|x| 
                x.compressed_account != compressed_account.compressed_account
            );
```

File: test-utils/src/test_indexer.rs [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+                .compressed_accounts
+                .iter()
+                .position(|x| x == compressed_account);
+            if let Some(index) = index {
+                let token_compressed_account_element =
+                    self.token_compressed_accounts[index].clone();
+                self.token_compressed_accounts.remove(index);
+                self.token_nullified_compressed_accounts
+                    .push(token_compressed_account_element);
+            }
```
Comment (vadorovsky) [inline test-utils/src/test_indexer.rs:189]:
```suggestion
            if let Some((index, _)) = self.compressed_accounts.iter().enumerate()
                    .find(|&(_, acc)| acc == compressed_account) {
                let token_compressed_account_element =
                    self.token_compressed_accounts.remove(index);
                self.token_nullified_compressed_accounts.push(
                    token_compressed_account_element
                );
            }
```

File: test-utils/src/test_indexer.rs (hunk 3 above)
Comment (ananas-block) [reply test-utils/src/test_indexer.rs:175]: interesting, makes sense here for simplicity.

File: examples/token-escrow/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+            "compilerOptions": {
+              "types": ["mocha", "chai"],
+              "typeRoots": ["./node_modules/@types"],
+              "lib": ["es2015"],
+              "module": "commonjs",
+              "target": "es6",
+              "esModuleInterop": true
+            }
+          }
+
```
Comment (SwenSchaeferjohann) [inline examples/token-escrow/tsconfig.json:11]: not formatted

File: js/compressed-token/src/idl/psp_compressed_token.ts [hunk 6]
```diff
@@ -144,12 +144,12 @@ export type PspCompressedToken = {
             accounts: [
                 {
                     name: 'feePayer';
-                    isMut: true;
+                    isMut: false;
```
Comment (SwenSchaeferjohann) [inline js/compressed-token/src/idl/psp_compressed_token.ts:147]: why does `isMut` for compressed-token change?

File: programs/compressed-token/src/process_transfer.rs [hunk 7]
```diff
@@ -481,7 +487,115 @@ pub mod transfer_sdk {
...
+        owner_if_delegate_is_signer: Option<Pubkey>,
+        output_compressed_accounts: &[TokenTransferOutputData],
+        root_indices: &[u16],
+        proof: &CompressedProof,
+        mint: Pubkey,
+        owner: &Pubkey,
+        is_compress: bool,
+        compression_amount: Option<u64>,
+        // token_pool_pda: Option<Pubkey>,
+        // decompress_token_account: Option<Pubkey>,
```
Comment (SwenSchaeferjohann) [inline programs/compressed-token/src/process_transfer.rs:555]: leftover ?

File: programs/compressed-token/src/process_transfer.rs [hunk 8]
```diff
@@ -481,7 +487,115 @@ pub mod transfer_sdk {
...
+            output_compressed_account_merkle_tree_pubkeys,
+            owner_if_delegate_is_signer,
+            output_compressed_accounts,
+            root_indices,
+            proof,
+            mint,
+            is_compress,
+            compression_amount,
+            // token_pool_pda,
+            // decompress_token_account,
```
Comment (SwenSchaeferjohann) [inline programs/compressed-token/src/process_transfer.rs:578]: leftover?

File: programs/compressed-token/src/process_transfer.rs [hunk 9]
```diff
@@ -481,7 +487,115 @@ pub mod transfer_sdk {
...
+        output_compressed_account_merkle_tree_pubkeys: &[Pubkey],
+        owner_if_delegate_is_signer: Option<Pubkey>,
+        output_compressed_accounts: &[TokenTransferOutputData],
+        root_indices: &[u16],
+        proof: &CompressedProof,
+        mint: Pubkey,
+        is_compress: bool,
+        compression_amount: Option<u64>,
+        // token_pool_pda: Option<Pubkey>,
+        // decompress_token_account: Option<Pubkey>,
```
Comment (SwenSchaeferjohann) [inline programs/compressed-token/src/process_transfer.rs:597]: leftover?

**Pull request merged**